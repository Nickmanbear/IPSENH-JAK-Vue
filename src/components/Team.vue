<template>
  <div id="team">
    <button v-if="!editingName && isLeader" id="delete" @click="deleteTeam">&times;</button>
    <h2 v-if="!editingName" @click="editingName = true">{{ team.name }}</h2>
    <input id="team-name" v-else v-model="team.name" @keydown.enter="saveTeam" type="text">
    <button v-if="editingName" @click="saveTeam">save</button>

    <div v-if="isLeader">
      <input v-model="selectedMemberName" type="text" placeholder="Username" list="user-list">
      <datalist id="user-list" v-if="selectedMemberName">
        <option v-for="user in filteredUsers" :key="user.id" :value="user.username"/>
      </datalist>
      <button @click="addMember">Add member</button>
    </div>

      <p v-for="member in team.members" :key="member.id"
         v-bind:class="{ memberName: isLeader }"
         @click="() => deleteMember(member.id)">{{ member.username }}</p>
  </div>
</template>

<script>
// @ is an alias to /src
import axios from '@/axiosInstance';

export default {
  name: 'Team',
  props: {
    team: {
      id: 0,
      name: '',
      leader: {},
      members: [],
    },
    users: {},
    username: String,
  },
  data() {
    return {
      isLeader: true,
      editingName: false,
      selectedMemberName: '',
    };
  },
  computed: {
    filteredUsers() {
      let temporaryUsers = this.users
        .map((user) => {
          if (!this.userInTeam(user)) {
            return user;
          }
          return null;
        });
      temporaryUsers = temporaryUsers.filter((user) => (user !== null));
      return temporaryUsers;
    },
  },
  mounted() {
    this.isLeader = this.team.leader.username === this.username;
  },
  methods: {
    userInTeam(user) {
      return this.team.members.find(
        (member) => member.username === user.username,
      );
    },
    saveTeam() {
      this.editingName = false;
      axios.post('/team', this.team)
        .then(() => {
          // TODO update
        });
    },
    addMember() {
      const selectedUser = this.users.find(
        (user) => user.username === this.selectedMemberName,
      );
      this.selectedMemberName = '';
      if (!selectedUser) {
        alert('This user doesn\'t exist.');
        return;
      }

      axios.post(`/team/member/${this.team.id}/${selectedUser.id}`)
        .then(() => {
          this.$emit('refresh');
        });
    },
    deleteTeam() {
      axios.delete(`/team/${this.team.id}`)
        .then(() => {
          this.$emit('refresh');
        });
    },
    deleteMember(userId) {
      axios.delete(`/team/member/${this.team.id}/${userId}`)
        .then(() => {
          this.$emit('refresh');
        });
    },
  },
};
</script>

<style lang="scss">
  #team {
    input {
      margin: 0;
    }

    h2 {
      margin-bottom: 0;
    }

    p {
      margin: 5px 0 0 0;
    }

    #delete {
      font-size: 1em;
      border: none;
      background-color: transparent;
      cursor: pointer;
      padding: 5px 10px;
      position: relative;
      float: right;
      border-radius: 50%;
      color: #fff;
      top: 0;
      transition: all 0.3s ease-out;

      &:hover {
        color: red;
        background-color: #f4f4f4;
      }
    }

    .memberName:hover {
      text-decoration: line-through;
      cursor: pointer;
    }
  }
</style>
