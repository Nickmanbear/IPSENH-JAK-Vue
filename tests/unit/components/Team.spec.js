import { shallowMount } from '@vue/test-utils';
import Team from '@/components/Team.vue';

const $route = { params: { id: 1 } };

jest.mock('axios', () => ({
  create: () => ({
    post: () => Promise.resolve(null),
    delete: () => Promise.resolve(null),
  }),
}));

describe('Team', () => {
  let cmp;

  beforeEach(() => {
    cmp = shallowMount(Team, {
      mocks: {
        $route,
      },
      propsData: {
        team: {
          id: 1,
          name: 'test team',
          leader: {
            id: 1,
            username: 'teamleader',
          },
          members: [
            {
              id: 1,
              username: 'teamleader',
            },
            {
              id: 2,
              username: 'member1',
            },
            {
              id: 3,
              username: 'member2',
            },
          ],
        },
        users: [
          {
            id: 4,
            username: 'user1',
          },
          {
            id: 5,
            username: 'user2',
          },
          {
            id: 6,
            username: 'user3',
          },
        ],
        username: 'teamleader',
      },
      data() {
        return {
          editingName: false,
          selectedMemberName: '',
        };
      },
    });
    window.alert = jest.fn(() => true);
  });

  it('should find user', () => {
    const user = { id: 3, username: 'member2' };
    expect(cmp.vm.userInTeam(user)).toEqual(user);
  });

  it('should not find user', () => {
    const user = { id: 4, username: 'user1' };
    expect(cmp.vm.userInTeam(user)).toEqual(undefined);
  });

  it('should update selectedMemberName', () => {
    cmp.find('input').setValue('test member name');
    expect(cmp.vm.selectedMemberName).toEqual('test member name');
  });

  it('should refresh on saveTeam', async () => {
    await cmp.vm.saveTeam();
    expect(cmp.vm.team).toEqual({
      id: 1,
      name: 'test team',
      leader: {
        id: 1,
        username: 'teamleader',
      },
      members: [
        {
          id: 1,
          username: 'teamleader',
        },
        {
          id: 2,
          username: 'member1',
        },
        {
          id: 3,
          username: 'member2',
        },
      ],
    });
  });

  it('should refresh on addMember', async () => {
    await cmp.vm.addMember();
    expect(cmp.vm.team).toEqual({
      id: 1,
      name: 'test team',
      leader: {
        id: 1,
        username: 'teamleader',
      },
      members: [
        {
          id: 1,
          username: 'teamleader',
        },
        {
          id: 2,
          username: 'member1',
        },
        {
          id: 3,
          username: 'member2',
        },
      ],
    });
  });

  it('should refresh on deleteTeam', async () => {
    await cmp.vm.deleteTeam();
    expect(cmp.vm.team).toEqual({
      id: 1,
      name: 'test team',
      leader: {
        id: 1,
        username: 'teamleader',
      },
      members: [
        {
          id: 1,
          username: 'teamleader',
        },
        {
          id: 2,
          username: 'member1',
        },
        {
          id: 3,
          username: 'member2',
        },
      ],
    });
  });

  it('should refresh on deleteMember', async () => {
    await cmp.vm.deleteMember(3);
    expect(cmp.vm.team).toEqual({
      id: 1,
      name: 'test team',
      leader: {
        id: 1,
        username: 'teamleader',
      },
      members: [
        {
          id: 1,
          username: 'teamleader',
        },
        {
          id: 2,
          username: 'member1',
        },
        {
          id: 3,
          username: 'member2',
        },
      ],
    });
  });

  it('has the expected html structure', () => {
    expect(cmp.element).toMatchSnapshot();
  });
});
