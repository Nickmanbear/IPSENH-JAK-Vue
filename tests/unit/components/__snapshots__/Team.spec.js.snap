// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team has the expected html structure 1`] = `
<div
  id="team"
>
  <button
    id="delete"
  >
    Ã—
  </button>
   
  <h2>
    test team
  </h2>
   
  <!---->
   
  <div>
    <input
      list="user-list"
      placeholder="Username"
      type="text"
    />
     
    <!---->
     
    <button>
      Add member
    </button>
  </div>
   
  <p
    class="memberName"
  >
    teamleader
  </p>
  <p
    class="memberName"
  >
    member1
  </p>
  <p
    class="memberName"
  >
    member2
  </p>
</div>
`;
